src/routing_table/prefix.rs

```diff
@@ -213,83 +282,99 @@ impl Prefix {
         name
     }
 
-    /// Returns the same prefix, with the last bit flipped, or unchanged, if empty.
-    pub fn sibling(&self) -> Prefix {
-        if self.bit_count > 0 {
-            self.with_flipped_bit((self.bit_count - 1) as usize)
-        } else {
-            *self
-        }
+    /// Returns the number of common leading bits with the input name, capped with prefix length.
+    pub fn common_prefix(&self, name: &XorName) -> usize {
+        cmp::min(self.bit_count(), self.name.common_prefix(name))
     }
 }
 
-impl PartialEq<Prefix> for Prefix {
+impl PartialEq<Self> for UnversionedPrefix {
     fn eq(&self, other: &Self) -> bool {
         self.is_compatible(other) && self.bit_count == other.bit_count
     }
 }
 
-impl PartialOrd<Prefix> for Prefix {
+impl PartialOrd<Self> for UnversionedPrefix {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl Ord for Prefix {
+impl Ord for UnversionedPrefix {
     fn cmp(&self, other: &Self) -> Ordering {
-        if self == other {
-            Ordering::Equal
-        } else if self.is_compatible(other) {
+        if self.is_compatible(other) {
             self.bit_count().cmp(&other.bit_count())
         } else {
             self.name.cmp(&other.name)
         }
     }
 }
 
-impl Hash for Prefix {
+impl Hash for UnversionedPrefix {
     fn hash<H: Hasher>(&self, state: &mut H) {
         for i in 0..self.bit_count() {
             self.name.bit(i).hash(state);
         }
     }
 }
 
-impl Binary for Prefix {
+impl Binary for UnversionedPrefix {
     fn fmt(&self, formatter: &mut Formatter) -> FmtResult {
         let mut binary = self.name.binary();
         binary.truncate(self.bit_count());
         write!(formatter, "{}", binary)
     }
 }
 
-impl Debug for Prefix {
+impl Debug for UnversionedPrefix {
     fn fmt(&self, formatter: &mut Formatter) -> FmtResult {
         write!(formatter, "Prefix({:b})", self)
     }
 }
 
-#[cfg(test)]
-impl FromStr for Prefix {
-    type Err = String;
-    fn from_str(bits: &str) -> Result<Prefix, String> {
-        let mut first_byte = 0u8;
-        for (i, bit) in bits.chars().enumerate() {
-            if bit == '1' {
-                first_byte |= 1 << (7 - i);
-            } else if bit != '0' {
-                return Err(format!(
-                    "'{}' not allowed - the string must represent a binary number.",
-                    bit
-                ));
-            }
+/// Find the entry at the given prefix, ignoring versions.
+pub fn unversioned_find<'a, T>(
+    map: &'a BTreeMap<Prefix, T>,
+    key: &UnversionedPrefix,
+) -> Option<(Prefix, &'a T)> {
+    map.range(key.with_version(0)..key.with_version(u64::MAX))
+        .next()
+        .map(|(prefix, value)| (*prefix, value))
+}
```

Comment:
just wondering this `unversioned` methods actually indicates the map shall having `UnversionedPrefix` as key?
otherwise, the insertion into map allow same prefix having different version and the `unversioned_get/find` methods only return the first matched one.
maybe we shall have another `unversion_insert` and ensure the map calls unversion_insert as well?

Reply:
The purpose of this method is to allow you to have a map keyed by **versioned** `Prefix`es, but to query it by `UnversionedPrefix`es.

`unversioned_insert` wouldn't make sense in my opinion, because we need the keys to be versioned. It's currently up to the user of the map to make sure no prefix is present in the map more than once with just the version number differing. I believe `RoutingTable` currently takes care of that. We could in theory write a custom collection for this, but it might be overkill. Not sure.

Reply:
I don't see  `RoutingTable` carry out any explicit check to ensure that  no prefix is present in the map more than once with just the version number differing.
It's more like totally depends on the events that RoutingTable receives.

another way around, shall those methods using unversioned prefix shall have `Prefix` as input parameter?

Reply:
I think that's handled by `insert_new_section()`, but it may be worth ensuring that the invariant check fails if this case ever happens?

Reply:
Hm, I think I misunderstood what you meant about the `unversioned_insert` function, @maqi . Did you mean something like this:

```rust
fn unversioned_insert<T>(map: &mut BTreeMap<Prefix, T>, key: Prefix, value: T) -> Option<T> {
    let result = unversioned_remove(map, &key.unversioned());
    let _ = map.insert(key, value);
    result
}
```
? Because I think that might be actually useful.

Reply:
yes.

I will prefer `not to carry out insertion if there is prefix having diff version exists`.
But not strongly on it.